cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/KafkaCruiseControlServlet.java

```diff
@@ -404,26 +489,52 @@ private void getProposals(HttpServletRequest request, HttpServletResponse respon
+    String loadBeforeOptimization;
+    String loadAfterOptimization;
+    if (json) {
+      loadBeforeOptimization = optimizerResult.brokerStatsBeforeOptimization().getJSONString();
+      loadAfterOptimization = optimizerResult.brokerStatsAfterOptimization().getJSONString();
     } else {
-      out.write(optimizerResult.goalProposals().toString().getBytes(StandardCharsets.UTF_8));
+      loadBeforeOptimization = optimizerResult.brokerStatsBeforeOptimization().toString();
+      loadAfterOptimization = optimizerResult.brokerStatsAfterOptimization().toString();
     }
-    for (Map.Entry<Goal, ClusterModelStats> entry : optimizerResult.statsByGoalPriority().entrySet()) {
-      Goal goal = entry.getKey();
-      out.write(String.format("%n%nStats for goal %s%s:%n", goal.name(), goalResultDescription(goal, optimizerResult))
-                      .getBytes(StandardCharsets.UTF_8));
-      out.write(entry.getValue().toString().getBytes(StandardCharsets.UTF_8));
+    OutputStream out = response.getOutputStream();
+    if (!json) {
+      if (!verbose) {
+        out.write(KafkaCruiseControlServletUtils.getProposalSummary(optimizerResult).getBytes(StandardCharsets.UTF_8));
+      } else {
+        out.write(optimizerResult.goalProposals().toString().getBytes(StandardCharsets.UTF_8));
+      }
+      setResponseCode(response, HttpServletResponse.SC_OK);
+      // Print summary before & after optimization
+      out.write(String.format("%n%nCurrent load:").getBytes(StandardCharsets.UTF_8));
+      out.write(loadBeforeOptimization.getBytes(StandardCharsets.UTF_8));
+      out.write(String.format("%n%nOptimized load:").getBytes(StandardCharsets.UTF_8));
+      out.write(loadAfterOptimization.getBytes(StandardCharsets.UTF_8));
+    } else {
+      // Build all the goal summary
+      List<Map<String, Object>> allGoals = new ArrayList<>();
+      for (Map.Entry<Goal, ClusterModelStats> entry : optimizerResult.statsByGoalPriority().entrySet()) {
+        Goal goal = entry.getKey();
+        String goalViolation = goalResultDescription(goal, optimizerResult);
+        Map<String, Object> goalMap = new HashMap<>();
+        goalMap.put("goal", goal.name());
+        goalMap.put("goalViolated", goalViolation);
+        goalMap.put("clusterModelStats", entry.getValue().getJsonStructure());
+        allGoals.add(goalMap);
+      }
+      Map<String, Object> proposalMap = new HashMap<>();
+      proposalMap.put("goals", allGoals);
+      proposalMap.put("loadBeforeOptimization", optimizerResult.brokerStatsBeforeOptimization().getJsonStructure());
+      proposalMap.put("loadAfterOptimization", optimizerResult.brokerStatsAfterOptimization().getJsonStructure());
+      Gson gson = new GsonBuilder()
+                      .serializeNulls()
+                      .serializeSpecialFloatingPointValues()
+                      .create();
+      String proposalsString = gson.toJson(proposalMap);
```

Comment:
Ideally we should stream the result to the socket instead of form a big string for performance purpose.

Reply:
same as above